pep-9999.rst

```diff
@@ -0,0 +1,731 @@
+
+-  Any: e.g. ``Literal[Any]`` Note: the semantics of what exactly
+   ``Literal[Any]`` means would need to be clarified first.
+
+Parameters at runtime
+---------------------
+
+Although the set of parameters ``Literal[...]`` may contain at type-check time
+is very small, the actual implementation of ``typing.Literal`` will not perform
+any checks at runtime. For example::
+
+   def my_function(x: Literal[1 + 2]) -> int:
+       return x * 3
+       
+   x: Literal = 3
+   y: Literal[my_function] = my_function
+
+The type checker should reject this program: all three uses of
+``Literal`` are *invalid* according to this spec. However, Python itself
+should execute this program with no errors.
+
+This is partly to help us preserve flexibility in case we want to expand the
+scope of what ``Literal`` can be used for in the future, and partly because
+it is not possible to detect all illegal parameters at runtime to begin with.
+For example, it is impossible to distinguish between ``Literal[1 + 2]`` and 
+``Literal[3]`` at runtime.
+
+Literals, enums, and forward references
+---------------------------------------
+
+One potential ambiguity is between literal strings and forward
+references to literal enum members. For example, suppose we have the
+type ``Literal["Color.RED"]``. Does this literal type
+contain a string literal or a forward reference to some ``Color.RED``
+enum member?
+
+In cases like these, we always assume the user meant to construct a
+literal string. If the user wants a forward reference, they must wrap
+the entire literal type in a string -- e.g. ``"Literal[Color.RED]"``.
+
+Literals, enums, and Any
+------------------------
+
+Another ambiguity is when the user attempts to use some expression that
+is meant to be an enum but is actually of type ``Any``. For example,
+suppose a user attempts to import an enum from a package with no type hints::
+
+   from typing import Literal
+   from lib_with_no_types import SomeEnum  # SomeEnum has type 'Any'!
+
+   # x has type `Literal[Any]` due to the bad import
+   x: Literal[SomeEnum.FOO]
+
+Because ``Literal`` may not be parameterized by ``Any``, this program
+is *illegal*: the type checker should report an error with the last line.
+
+In short, while ``Any`` may effectively be used as a placeholder for any
+arbitrary *type*, it is currently **not** allowed to serve as a placeholder
+for any arbitrary *value*.
+
+
+Inferring types for literal expressions
+=======================================
+
+This section describes under what circumstances some expression should have
+an inferred Literal type.
+
+For example, under what circumstances should literal expressions like ``"blue"``
+have an inferred type of ``Literal["blue"]`` vs ``str``?
+
+In short, type checkers are expected to be conservative and bias towards
+inferring standard types like ``str``. Type checkers should infer ``Literal[...]``
+only in contexts where a Literal type is explicitly requested.
```

Comment:
What I was trying to say at the meetup is that this kind of inference guidance is too prescriptive in my opinion.  As far as I know, inference has so far been unspecified in the typing PEPs, and I think it's better that way, as mypy and Pyre seem to have different behavior.

The key example I'd like to be supported, but not required behavior is

```
x = 5
x = 6
takes_six(x)
```

This currently goes against lines like 
>When assigning a literal expression to an unannotated variable, the inferred type of the variable is the original base type, not ``Literal[...]``.


Pyre allows variables without explicit annotations to have different types at different points in a function, and we'd like to maintain this flexibility for literals.

Ultimately, I think Note 2 is more of the spirit of what should be in the PEP, and most of this Inferring types for literal expressions section can and should be removed

Reply:
@mrkmndz -- I'm thinking of replacing these sections with something roughly like this:

 > When type checkers are implementing `Literal`, it's important they do so in a way that preserves backwards compatibility. Code that currently type-checks should continue to type-check after adding support for Literal.
>
> For example, take a statement like `x = 6`. It would be incorrect for type-checkers to automatically assume `x` is of type `Literal[6]` since that would break code that looks like this:
>
>     def expects_int(x: int) -> None: ...
>     x = 6
>     expects_int(6)  # Previously type-checked, and so should continue to type-check
>
>
> In cases like these, type checkers should either perform type inference to determine the correct type of `x`, or if they are unable to do so, conservatively bias towards assuming that `x` is of type `int` instead of a Literal type.
>
> [Add more examples and edge cases of what type checkers shouldn't do here]

...and like you suggested, remove any text that requires a specific inference strategy.

Does this seem good to you? If so, I can make this change later today + ping you once that's ready.

Reply:
What inference would lead to that example not type checking?  If I understand correctly, even with an explicit annotation of Literal[6], that example should still type check as literal integers are subtypes of int?

Reply:
Oh whoops, you're right -- that's a terrible example. (Sorry, was rushing
to get the comment out before lunch).

Something like:

    x = 6
    x = 100

...would be a better example: we can't infer x is of type `Literal[6]` there.

Reply:
Or, you could, so long as you support changing the type of x -- that would also preserve backwards compatibility.